dfm/data/deduper.py

```diff
@@ -0,0 +1,233 @@
+        num_minhashes (int, optional):
+            The number of MinHash functions to use. Defaults to 128.
+        random_seed (int, optional):
+            The random seed to use for the MinHash functions. Defaults to 42.
+
+    Attributes:
+        split_method (str): The splitting method for extracting shingles.
+        ngram_size (str): The size of the ngram shingles.
+        ngram_stride (str): The stride used for the ngram shingles.
+        similarity_threshold (float): The Jaccard similarity threshold.
+        num_minhashes (int): The number of MinHash functions to use.
+        random_seed (int): The random seed to use for the MinHash functions.
+        cache (MinHashLSH object): The LSH cache for storing the fingerprints.
+
+    References:
+        [1] Broder, Andrei Z. "On the resemblance and containment of documents."
+            Proceedings. Compression and Complexity of SEQUENCES 1997
+            (Cat. No. 97TB100171). IEEE, 1997.
+    """
+
+    def __init__(
+        self,
+        split_method: Optional[str] = "char_ngram",
+        ngram_size: int = 13,
+        ngram_stride: int = 1,
+        similarity_threshold: float = 0.8,
+        num_minhashes: int = 128,
+        random_seed: int = 42,
+    ):
+        self.split_method = "none" if split_method is None else split_method
+        self.ngram_size = ngram_size
+        self.ngram_stride = ngram_stride
+        self.similarity_threshold = similarity_threshold
+        self.num_minhashes = num_minhashes
+        self.random_seed = random_seed
+
+    def _get_minhash(self, doc: str) -> LeanMinHash:
+        """Returns a minhash fingerprint for the given document.
+
+        Args:
+            doc (str): The document to create the MinHash object for.
+
+        Returns:
+            LeanMinHash: The minhash fingerprint for the given document.
+
+        Raises:
+            ValueError:
+                If `self.split_method` is not 'char_ngram', 'word_ngram',
+                'paragraph' or 'none'.
+        """
+        # NFKC normalise document and remove punctuation
+        doc = normalize("NFKC", doc)
+        doc = re.sub(r"[\.\,\:\;\!\?\(\)\[\]\{\}]", " ", doc)
+        doc = re.sub(" +", " ", doc)
+
+        # Initialise the fingerprint
+        minhash = MinHash(num_perm=self.num_minhashes, seed=self.random_seed)
+
+        # Extract shingles from the document, depending on the `split_method`
+        if self.split_method == "char_ngram":
+            max_char_idx = len(doc) - self.ngram_size
+            shingles = [
+                doc[i : i + self.ngram_size]
+                for i in range(0, max_char_idx, self.ngram_stride)
+            ]
+        elif self.split_method == "word_ngram":
+            words = doc.split(" ")
+            max_word_idx = len(words) - self.ngram_size
+            shingles = [
+                " ".join(words[i : i + self.ngram_size]).strip()
+                for i in range(0, max_word_idx, self.ngram_stride)
+            ]
```

Comment:
We should probably filter out empty strings here, so:

```suggestion
            shingles = list(filter(lambda x: x,
                (" ".join(words[i : i + self.ngram_size]).strip()
                for i in range(0, max_word_idx, self.ngram_stride))
            ))
            
```

<img width="474" alt="Screenshot 2022-01-29 at 21 57 05" src="https://user-images.githubusercontent.com/23721977/151677312-e1041a58-3a10-487a-bbb6-80d65f89147d.png">

Reply:
Good point. I tried to keep things consistent and fixed it in the definition of `words` instead.